src/API.jl

```diff
@@ -569,27 +569,47 @@ function status(ctx::Context, mode=PKGMODE_PROJECT)
     return
 end
 
-activate() = (Base.ACTIVE_PROJECT[] = Base.load_path_expand(nothing))
-function activate(path::String)
-    devpath = nothing
-    env = Base.active_project() === nothing ? nothing : EnvCache()
-    if env !== nothing && haskey(env.project["deps"], path)
-        uuid = UUID(env.project["deps"][path])
-        info = manifest_info(env, uuid)
-        devpath = haskey(info, "path") ? joinpath(dirname(env.project_file), info["path"]) : nothing
-    end
-    # `pkg> activate path`/`Pkg.activate(path)` does the following
-    # 1. if path exists, activate that
-    # 2. if path exists in deps, and the dep is deved, activate that path (`devpath` above)
-    # 3. activate the non-existing directory (e.g. as in `pkg> activate .` for initing a new env)
-    if Types.isdir_windows_workaround(path)
-        path = abspath(path)
-    elseif devpath !== nothing
-        path = abspath(devpath)
+activate() = (Base.ACTIVE_PROJECT[] = nothing)
+function activate(path::String; shared::Bool=false)
+    if !shared
+        devpath = nothing
+        env = Base.active_project() === nothing ? nothing : EnvCache()
+        if env !== nothing && haskey(env.project["deps"], path)
+            uuid = UUID(env.project["deps"][path])
+            info = manifest_info(env, uuid)
+            devpath = haskey(info, "path") ? joinpath(dirname(env.project_file), info["path"]) : nothing
+        end
+        # `pkg> activate path`/`Pkg.activate(path)` does the following
+        # 1. if path exists, activate that
+        # 2. if path exists in deps, and the dep is deved, activate that path (`devpath` above)
+        # 3. activate the non-existing directory (e.g. as in `pkg> activate .` for initing a new env)
+        if Types.isdir_windows_workaround(path)
+            fullpath = abspath(path)
+        elseif devpath !== nothing
+            fullpath = abspath(devpath)
+        else
+            fullpath = abspath(path)
+            isdir(fullpath) || @info("new environment will be placed at $fullpath")
+        end
     else
-        path = abspath(path)
+        # initialize `fullpath` in case of empty `Pkg.depots()`
+        fullpath = ""
+        # loop over all depots to check if the shared environment already exists
+        for depot in Pkg.depots()
+            fullpath = joinpath(Pkg.envdir(depot), path)
+            isdir(fullpath) && break
+        end
+        # this disallows names such as "Foo/bar", ".", "..", etc
+        if basename(abspath(joinpath(fullpath,path))) != path
```

Comment:
oh right thats an embarrassing mistake. also good point about line 608. Thanks!

Reply:
Welcome!

Reply:
moving the line to 608 would allow the user to say `activate --shared .` (which would resolve to the `environments` folder itself, as well as any `../../xx` to some existing folder. Unsure if it makes sense to allow or disallow that. At the very least `activate --shared .` seems like something that no one would every conciously want to do, so this might be a good case where we should rather throw some error